design/one-pager-trait-workload-interaction-mechanism.md

```diff
@@ -0,0 +1,198 @@
+`trait` CR. The application controller actually generates the workload CR so it has all the
+information. In our example, the applicationConfig controller will emit the following `trait` CR with
+the GVK and name of the workload the `ManualScalerTrait` is applied to.
+    ```yaml
+    apiVersion: core.oam.dev/v1alpha2
+    kind: ManualScalerTrait
+    metadata:
+      annotations:
+        core.oam.dev/workloadGroupVersion: standard.oam.dev/v1alpha2
+        core.oam.dev/workloadKind: Mydb
+        core.oam.dev/workloadName: mydb-example
+    ```
+4. OAM runtime will provide a helper library. The library follows the following logic to help a
+ trait developer locate the resources for the trait to modify
+    1. Get the corresponding `workload` instance from the Kubernetes cluster with the information
+     contained in the annotation field inserted by the application controller in the `trait` CR.
+    2. Fetch all the `childResources` values inserted by the application controller in the
+     annotation field in the `workload` instance. 
+    3. List each child resource by its GVK and filter by owner reference. Here, we assume that
+     all the child resources that the workload controller generates have an owner reference field
+      pointing back to the workload instance.
+
+## Impact to the existing system
+Here are the impact of this mechanism to the existing controllers
+- ApplicationConfiguration: This mechanism requires changes in the
+ applicationConfiguration controller.
+- Workload: This mechanism does not affect workload implementation.
+- Trait: This mechanism is optional so all existing trait controller still works. This mechanism
+requires modification for any existing traits implementation that wants to take advantage of
+extensibility of OAM. Any traits that only applies to a certain type of workload don't need to
+use this mechanism.
+- WorkloadDefinition: We can modify the existing workloadDefinition if needed.
+
+## Alternative approach
+1. One alternative approach is that we can make the applicationConfiguration controller watch all
+the possible workload child resources. It also inserts the child resources GVK and name to the
+correspoinding workload CR. I would not recommend this approach as it increases the complexity
+of the applicationConfiguration controller and makes more of availability liability. 
+2. Another approach is to implement a separate type for binding traits to workloads. This would
+ work but it seems that label/annotation is a natural place to record the information. Otherwise
+ , we need a way for the trait to discover the binding instance first.   
+
+## Extra labels
+There might be cases that a workload generates more than one resources with the same GVK and only
+want to expose a subset of them to traits. In this case, we can add a pre-defined label such as
+`core.oam.dev/expose=true` for the workload owner to indicate what resources to expose. This
```

Comment:
Ah nice! This addresses my comment above about how a trait will modify multiple resources of the same kind rendered by a workload. I still wonder about listing a resource multiple times in the `childResources` section. Maybe this is not necessary though as the trait reconciler should theoretically list all of every type of resource it cares about with owner references to the workload it cares about 👍